### PR TITLE
Godriver2259 Reduce memory consumption in Connection.CompressWireMessage().

### PR DESCRIPTION
**Pull request by qingyang-hu**
[GODRIVER-2259](https://jira.mongodb.org/browse/GODRIVER-2259)
[GODRIVER-2023](https://jira.mongodb.org/browse/GODRIVER-2023)

- Tune the Zstd compression.
- Use src slice directly when using no compression.

### CODE REVIEW

File: x/mongo/driver/topology/connection.go [hunk 1]
```diff
@@ -638,6 +638,9 @@ func (c *Connection) CompressWireMessage(src, dst []byte) ([]byte, error) {
 		return dst, ErrConnectionClosed
 	}
 	if c.connection.compressor == wiremessage.CompressorNoOp {
+		if len(dst) == 0 {
+			return src, nil
```
Comment (benjirewis) [inline x/mongo/driver/topology/connection.go:642]: Is there any case where `len(dst) > 0`? I think this is a good safeguard to have, but I'm not sure I can think of a case where the passed-in destination slice is non-empty. In any case, can we leave a comment above this line explaining why we can just return `src` in this case?

File: x/mongo/driver/compression.go [hunk 2]
```diff
@@ -25,6 +25,32 @@ type CompressionOpts struct {
...
+			windowSize = 16 << 20
+		case zstd.SpeedBestCompression:
+			windowSize = 32 << 20
+		}
+		if windowSize > zstd.MaxWindowSize {
+			windowSize = zstd.MaxWindowSize
+		}
+		// Reduce the window size to the closest power of 2 that can hold the input size
+		// if the default window size is larger than the input size.
+		for windowSize/2 > n {
```
Comment (benjirewis) [inline x/mongo/driver/compression.go:47]: So it looks like this reduction is our custom piece of window size calculation, and using the smallest window size possible results in less memory being allocated? Could you point me to the analogue of `calcZstdWindowSize` in the `zstd` library?

File: x/mongo/driver/compression_test.go [hunk 3]
```diff
@@ -40,3 +43,78 @@ func TestCompression(t *testing.T) {
...
+
+func TestZstdWindowSize(t *testing.T) {
+	tests := []struct {
+		inputSize  int
+		level      zstd.EncoderLevel
+		windowSize int
+	}{
+		{
+			inputSize:  512,
+			level:      zstd.EncoderLevelFromZstd(wiremessage.DefaultZstdLevel),
```
Comment (benjirewis) [inline x/mongo/driver/compression_test.go:55]: Could we add tests with different `EncoderLevel`s?

File: x/mongo/driver/compression_test.go [hunk 4]
```diff
@@ -40,3 +43,78 @@ func TestCompression(t *testing.T) {
...
+		t.Run(strconv.Itoa(test.inputSize), func(t *testing.T) {
+			windowSize := calcZstdWindowSize(test.inputSize, test.level)
+			assert.Equal(t, test.windowSize, windowSize)
+		})
+	}
+}
+
+func BenchmarkCompression(b *testing.B) {
+	payload := func() []byte {
+		f, err := os.Open("compression.go")
```
Comment (benjirewis) [inline x/mongo/driver/compression_test.go:85]: Are you compressing `compression.go` as the benchmark? Cool 😆

File: x/mongo/driver/compression_test.go [hunk 5]
```diff
@@ -40,3 +43,78 @@ func TestCompression(t *testing.T) {
...
+	}()
+
+	compressors := []wiremessage.CompressorID{
+		wiremessage.CompressorSnappy,
+		wiremessage.CompressorZLib,
+		wiremessage.CompressorZstd,
+	}
+
+	for _, compressor := range compressors {
+		b.Run(strconv.Itoa(int(compressor)), func(b *testing.B) {
```
Comment (benjirewis) [inline x/mongo/driver/compression_test.go:106]: [nit] Looks like the benchmark name will just be the number of the compressor in `iota` order. Could we use the actual names of the compressors instead (might have to make this table-driven).

File: x/mongo/driver/topology/connection_test.go [hunk 6]
```diff
@@ -926,6 +928,23 @@ func TestConnection(t *testing.T) {
 	})
 }
 
+func BenchmarkConnection(b *testing.B) {
+	b.Run("CompressWireMessage CompressorNoOp", func(b *testing.B) {
+		buf := make([]byte, 256)
+		_, err := rand.Read(buf)
```
Comment (benjirewis) [inline x/mongo/driver/topology/connection_test.go:934]: [nit] If we fill this buffer randomly on every call to `BenchmarkConnection` won't this have inconsistent results? I forget how this `rand` is seeded, though.

**Review by benjirewis — COMMENTED** [review]
Looking great so far! Just a few comments/questions.

File: x/mongo/driver/compression.go (hunk 2 above)
Comment (qingyang-hu) [reply x/mongo/driver/compression.go:47]: Sorry, I forgot to leave a reference.
The original code is [here](https://github.com/klauspost/compress/blob/master/zstd/encoder_options.go#L215).
And the instruction on the window size setting can be found [here](https://pkg.go.dev/github.com/klauspost/compress/zstd#WithWindowSize).

File: x/mongo/driver/compression_test.go (hunk 4 above)
Comment (qingyang-hu) [reply x/mongo/driver/compression_test.go:85]: I was trying to compress random bytes, but the compressor did poorly on them with compression ratio > 1. Also, compression.go is better than [Lorem ipsum](https://en.wikipedia.org/wiki/Lorem_ipsum), which I have to save the text somewhere else.

File: x/mongo/driver/topology/connection_test.go (hunk 6 above)
Comment (qingyang-hu) [reply x/mongo/driver/topology/connection_test.go:934]: Mostly, this benchmark is for `CompressorNoOp` only, so its content does not matter very much. The actual benchmark starts at l939 underneath.

File: x/mongo/driver/compression.go [hunk 7]
```diff
@@ -25,6 +25,32 @@ type CompressionOpts struct {
 	UncompressedSize int32
 }
 
+func calcZstdWindowSize(n int, l zstd.EncoderLevel) int {
+	windowSize := zstd.MinWindowSize
+	if n > windowSize {
```
Comment (prestonvasquez) [inline x/mongo/driver/compression.go:30]: What are your thoughts on removing this branch altogether?  One edge case I'm thinking of is if this routine is called with an input size of 0, i.e. this test case:

```
{
	inputSize:  0,
	level:      zstd.EncoderLevelFromZstd(wiremessage.DefaultZstdLevel),
	windowSize: 1,
},
```
Perhaps I'm not thinking about it correctly, but I would expect the window size to be 1 since that is the [smallest possible window size](https://github.com/klauspost/compress/blob/v1.15.6/zstd/framedec.go#L37).  Removing the `if` block fixes this problem with this block:

```go
for windowSize/2 > n {
	windowSize /= 2
}
```

**Review by prestonvasquez — COMMENTED** [review]
I love this solution!

File: x/mongo/driver/topology/connection.go (hunk 1 above)
Comment (matthewdale) [reply x/mongo/driver/topology/connection.go:642]: I looked into how the `CompressWireMessage` function is used and into where the `driver.Compressor` interface was introduced. In the driver, `CompressWireMessage` is always called with `dst=nil`, so the `dst` argument to `CompressWireMessage` seems unnecessary. However, it is technically possible to provide a custom dialer that returns connections that satisfy the driver.Compressor` interface. If we changed the `driver.Compressor` interface, we could technically break those implementations.

Consider the example:
```go
type myConn struct {
	net.Conn
}

func (c *myConn) CompressWireMessage(src, dst []byte) ([]byte, error) {
	// ...
}

type myDialer struct {}

func (d *myDialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
	return &myConn{}, nil
}

func main() {
	client, _ := mongo.Connect(context.Background() options.Client().SetDialer(&myDialer{}))
	// ...
}
```
`client` will now return connections with underlying type `*myConn`, which has a custom implementation of the `driver.Compressor` interface. It seems unlikely that anyone is writing their own compressor functions, but it's not clear if this would be considered a backward-breaking change or not. It wouldn't significantly simplify the `CompressWireMessage` function to remove the `dst` parameter, so I'd lean toward keeping the API the same to reduce the risk of breaking other implementations.

File: x/mongo/driver/compression.go [hunk 8]
```diff
@@ -25,6 +25,32 @@ type CompressionOpts struct {
...
+			windowSize = 32 << 20
+		}
+		if windowSize > zstd.MaxWindowSize {
+			windowSize = zstd.MaxWindowSize
+		}
+		// Reduce the window size to the closest power of 2 that can hold the input size
+		// if the default window size is larger than the input size.
+		for windowSize/2 > n {
+			windowSize /= 2
+		}
```
[the remark this answers is not in the document]
Comment (qingyang-hu) [reply x/mongo/driver/compression.go:49]: A loop of dividing 2 is more efficient than log2. A benchmark shows log2 takes ~4x time per op compared with looping. Log2 may be more readable? I think the comment is enough for readability.

File: x/mongo/driver/compression.go (hunk 7 above)
Comment (qingyang-hu) [reply x/mongo/driver/compression.go:30]: The smallest possible window size is 1024 (1 << 10). I did not dig very deep to verify if we still need a minimal window size for 0 input. However, it will not be wrong to keep it. Also, I keep the const of `MinWindowSize` as a safeguard to prevent any future changes in the third-part package.

Otherwise, I agree with @matthewdale that an early return for anything smaller than the `MinWindowSize` may be more readable.

File: x/mongo/driver/compression.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (qingyang-hu) [reply x/mongo/driver/compression.go]: I think it's because both `math.Log2` and `math.Ceil` take float parameters.

File: x/mongo/driver/compression_test.go [hunk 9]
```diff
@@ -40,3 +42,120 @@ func TestCompression(t *testing.T) {
...
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("size %d level %v", test.inputSize, test.level), func(t *testing.T) {
+			windowSize := calcZstdWindowSize(test.inputSize, test.level)
+			assert.Equal(t, test.windowSize, windowSize)
+		})
+	}
+}
+
+func BenchmarkCompression(b *testing.B) {
```
Comment (matthewdale) [inline x/mongo/driver/compression_test.go:127]: Optional: Consider renaming this to `BenchmarkCompressPayload` because `CompressPayload` is the specific function that's being benchmarked.

File: x/mongo/driver/compression_test.go [hunk 10]
```diff
@@ -40,3 +42,120 @@ func TestCompression(t *testing.T) {
...
+
+	for _, compressor := range compressors {
+		b.Run(compressor.String(), func(b *testing.B) {
+			opts := CompressionOpts{
+				Compressor: compressor,
+				ZlibLevel:  wiremessage.DefaultZlibLevel,
+				ZstdLevel:  wiremessage.DefaultZstdLevel,
+			}
+			for i := 0; i < b.N; i++ {
+				_, err := CompressPayload(payload, opts)
```
Comment (matthewdale) [inline x/mongo/driver/compression_test.go:154]: `CompressPayload` is intended for use by many goroutines, so this benchmark may provide more representative results if written as a parallel benchmark.

E.g.
```go
for _, compressor := range compressors {
	b.Run(compressor.String(), func(b *testing.B) {
		opts := CompressionOpts{
			Compressor: compressor,
			ZlibLevel:  wiremessage.DefaultZlibLevel,
			ZstdLevel:  wiremessage.DefaultZstdLevel,
		}
		b.RunParallel(func(pb *testing.PB) {
			for pb.Next() {
				_, err := CompressPayload(payload, opts)
				if err != nil {
					b.Error(err)
				}
			}
		})
	}
}
```

File: x/mongo/driver/topology/connection.go [hunk 11]
```diff
@@ -638,7 +638,7 @@ func (c *Connection) CompressWireMessage(src, dst []byte) ([]byte, error) {
 		return dst, ErrConnectionClosed
 	}
 	if c.connection.compressor == wiremessage.CompressorNoOp {
-		return append(dst, src...), nil
+		return src, nil
```
Comment (matthewdale) [inline x/mongo/driver/topology/connection.go:641]: To prevent backward-breaking changes and confusing behavior in the future, we should still support the case where `dst` is not empty/nil.

E.g.
```go
if len(dst) == 0 {
	return src, nil
}
return append(dst, src...), nil
```

**Comment by qingyang-hu** [comment]
After the discussion with @matthewdale, I switched to [EncodeAll()](https://pkg.go.dev/github.com/klauspost/compress@v1.13.6/zstd#Encoder.EncodeAll) to reuse the encoder. Benchmarks show an even lower memory consumption in the concurrent scenario.

**Review by prestonvasquez — APPROVED** [review]
LGTM!  I tested the benchmark locally and no issues.

File: x/mongo/driver/compression.go [hunk 12]
```diff
@@ -25,6 +26,21 @@ type CompressionOpts struct {
 	UncompressedSize int32
 }
 
+var zstdEncoders = &sync.Map{}
```
Comment (benjirewis) [inline x/mongo/driver/compression.go:29]: Haven't seen a `sync.Map` in our production code before! I think it's normally used instead of a regular map and a mutex for the following reasons:

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

I think (1) is true in this case, as we seem to create an encoder for each encoder level and then "read" it (`Load`) many times.

File: x/mongo/driver/compression.go [hunk 13]
```diff
@@ -48,21 +64,11 @@ func CompressPayload(in []byte, opts CompressionOpts) ([]byte, error) {
...
 		if err != nil {
-			_ = w.Close()
 			return nil, err
 		}
-		err = w.Close()
-		if err != nil {
-			return nil, err
-		}
-		return b.Bytes(), nil
+		return encoder.EncodeAll(in, nil), nil
```
Comment (benjirewis) [inline x/mongo/driver/compression.go:71]: Much simpler!

**Review by benjirewis — APPROVED** [review]
Looks good! Great work 😄

**Review by matthewdale — APPROVED** [review]
Looks good 👍 

@qingyang-hu Can you post benchmark comparisons for `BenchmarkCompressPayload` and `BenchmarkConnection` before and after these improvements?

**Comment by qingyang-hu** [comment]
### Benchmark results from [BenchmarkCompressPayload](https://github.com/mongodb/mongo-go-driver/blob/5d7bd99b167dcb19b89fe617215eacd08393b0f3/x/mongo/driver/compression_test.go#L44):

- OLD:
> CompressorZstd-4                 249    4412763 ns/op 38611199 B/op      135 allocs/op

- NEW:
> CompressorZstd-4                   5166        265163 ns/op        2968 B/op           4 allocs/op

- DELTA with "-count=10" using [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat)
> name old time/op new time/op delta
CompressorZstd-4 4.38ms ± 1% 0.23ms ± 1% -94.72% (p=0.000 n=10+10)
name old alloc/op new alloc/op delta
CompressorZstd-4 38.6MB ± 0% 0.0MB ± 0% -99.99% (p=0.000 n=10+10)
name old allocs/op new allocs/op delta
CompressorZstd-4 135 ± 0% 4 ± 0% -97.04% (p=0.000 n=10+10)

### Benchmark results from [BenchmarkConnection](https://github.com/mongodb/mongo-go-driver/blob/5d7bd99b167dcb19b89fe617215eacd08393b0f3/x/mongo/driver/topology/connection_test.go#L931):

- OLD:
> BenchmarkConnection/CompressWireMessage_CompressorNoOp-4 3836367 260.9 ns/op 256 B/op 1 allocs/op

- NEW:
> BenchmarkConnection/CompressWireMessage_CompressorNoOp-4 17769693 68.02 ns/op 0 B/op 0 allocs/op

- DELTA with "-count=10" using [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat)
> name old time/op new time/op delta
Connection/CompressWireMessage_CompressorNoOp-4 261ns ± 3% 48ns ±43% -81.69% (p=0.000 n=10+10)
name old alloc/op new alloc/op delta
Connection/CompressWireMessage_CompressorNoOp-4 256B ± 0% 0B -100.00% (p=0.000 n=10+10)
name old allocs/op new allocs/op delta
Connection/CompressWireMessage_CompressorNoOp-4 1.00 ± 0% 0.00 -100.00% (p=0.000 n=10+10)
